File: ruby solutions/linked_list_cycle.rb
# Definition for singly-linked list.
class ListNode
    attr_accessor :val, :next
    def initialize(val)
        @val = val
        @next = nil
    end
end

# @param {ListNode} head
# @return {Boolean}
require 'set'
def hasCycle(head)
    node_set = Set.new()
    while(head)
        if node_set.include? head
            return true
        end
        node_set.add(head)
        head = head.next
    end
    return false
end

File: python solutions/linked_list_cycle.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

from typing import Optional


class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        node_set = set()
        while(head):
            if head in node_set:
                return True
            node_set.add(head)
            head = head.next
        return False